robot/src/main/java/us/ilite/robot/modules/Elevator.java

```diff
@@ -0,0 +1,280 @@
+    private double mMinPower = -1.0d;
+    private double mHoldVoltage = 0.5; //NOT final
+    private boolean mSettingPosition = false;
+    private int kCansparkId = 15; //TODO change ID
+    private int kTalonId = 15; //TODO change ID
+    private PIDController mPidController;
+    private double mCurrentTime;
+
+    //TODO need to figure out these encoder thresholds
+    private int mDownEncoderThreshold = SystemSettings.kLowerElevatorEncoderThreshold;
+    private int mUpEncoderThreshold = SystemSettings.kLowerElevatorEncoderThreshold;
+
+    EElevatorPosition mCurrentPosition = EElevatorPosition.BOTTOM;
+
+    EElevatorState mCurrentState = EElevatorState.NORMAL;
+    EElevatorPosition mDesiredPosition;
+
+    CANSparkMax mMasterElevator;
+    TalonSRX mEncoder = TalonSRXFactory.createDefaultTalon(kTalonId);
+
+    public Elevator(Data pData) {
+
+        this.mData = pData;
+        PIDGains pidGains = new PIDGains( SystemSettings.kElevatorP, SystemSettings.kElevatorI, SystemSettings.kElevatorD );
+        this.mPidController = new PIDController( pidGains, SystemSettings.kELevatorControlLoopPeriod );
+
+        //Create default NEO and set the ramp rate
+        mMasterElevator = SparkMaxFactory.createDefaultSparkMax(kCansparkId, MotorType.kBrushless);
+        mMasterElevator.setIdleMode(IdleMode.kBrake);
+        mMasterElevator.setRampRate(SystemSettings.kELevatorControlLoopPeriod);
+
+
+        //We start at the bottom
+        mAtBottom = true;
+        mAtTop = false;
+        mDesiredDirectionUp = true;
+
+        //Make sure the elevator is stopped upon initialization
+        mDesiredPosition = EElevatorPosition.BOTTOM;
+        mCurrentState = EElevatorState.STOP;
+
+        mCurrentEncoderTicks = 0;
+    }
+
+    public void shutdown(double pNow) {
+
+    }
+
+    public void modeInit(double pNow) {
+
+    }
+
+    public void periodicInput(double pNow) {
+
+    }
+
+    public void update(double pNow) {
+
+        mCurrentTime = pNow;
+
+        //Get encoder value from secondary Talon
+        mCurrentEncoderTicks = getEncoderPosition();
+        mDesiredDirectionUp = (mDesiredPower > 0);
+
+        updateElevator(pNow);
+        mDesiredPower = calculateDesiredPower(mCurrentState);
+        mDesiredPower *= 0.10d; //10% of the desired power
+
+        mMasterElevator.set(Util.limit(mDesiredPower, mMinPower, mMaxPower));
+
+    }
+
+    public void updateElevator(double pNow) {
+
+        //TODO test this
+        //Should override any automated control
+        if (Math.abs(mDesiredPower) ==  0d && !mAtBottom) { //Hold if input is absent and we aren't at the bottom
+            mCurrentState = EElevatorState.HOLD;
+        } else {
+            mCurrentState = EElevatorState.NORMAL;
+        }
+
+        if (mDesiredDirectionUp) {
+            shouldDecelerate(mCurrentEncoderTicks, mDesiredDirectionUp);
+        }
+
+    }
+
+    private void resetBottom() {
+        mCurrentEncoderTicks = 0;
+        zeroEncoder();
+    }
+
```

Comment:
This method should read the # of encoder ticks and assume it is the max; this method should be called when we hit an upper limit.  Regardless, it also makes a huge assumption that the elevator was zero'ed at the bottom first.  We should add comments for what this method does.